File: libs/language/src/lib/provider/BasicNodeProvider.ts
import { BlueNode, NodeDeserializer } from '../model';
import { PreloadedNodeProvider } from './PreloadedNodeProvider';
import { Preprocessor } from '../preprocess/Preprocessor';
import { NodeContentHandler } from './NodeContentHandler';
import { BlueIdCalculator, NodeToMapListOrValue, Nodes } from '../utils';
import { yamlBlueParse } from '../../utils/yamlBlue';
import { JsonBlueValue } from '../../schema';

export class BasicNodeProvider extends PreloadedNodeProvider {
  private blueIdToContentMap: Map<string, JsonBlueValue> = new Map();
  private blueIdToMultipleDocumentsMap: Map<string, boolean> = new Map();
  private preprocessor: (node: BlueNode) => BlueNode;

  constructor(nodes: BlueNode[] = []) {
    super();

    // Create default preprocessor
    const defaultPreprocessor = new Preprocessor({ nodeProvider: this });
    this.preprocessor = (node: BlueNode) =>
      defaultPreprocessor.preprocessWithDefaultBlue(node);

    // Process initial nodes
    nodes.forEach((node) => this.processNode(node));
  }

  private processNode(node: BlueNode): void {
    if (Nodes.hasItemsOnly(node)) {
      this.processNodeWithItems(node);
    } else {
      this.processSingleNode(node);
    }
  }

  private processSingleNode(node: BlueNode): void {
    const parsedContent = NodeContentHandler.parseAndCalculateBlueIdForNode(
      node,
      this.preprocessor
    );
    this.blueIdToContentMap.set(parsedContent.blueId, parsedContent.content);
    this.blueIdToMultipleDocumentsMap.set(
      parsedContent.blueId,
      parsedContent.isMultipleDocuments
    );
    const nodeName = node.getName();
    if (nodeName) {
      this.addToNameMap(nodeName, parsedContent.blueId);
    }
  }

  private processNodeWithItems(node: BlueNode): void {
    const items = node.getItems();
    if (!items) return;

    this.processNodeList(items);

    const parsedContent = NodeContentHandler.parseAndCalculateBlueIdForNodeList(
      items,
      this.preprocessor
    );
    this.blueIdToContentMap.set(parsedContent.blueId, parsedContent.content);
    this.blueIdToMultipleDocumentsMap.set(parsedContent.blueId, true);

    items.forEach((item, i) => {
      const nodeName = item.getName();
      if (nodeName) {
        this.addToNameMap(nodeName, `${parsedContent.blueId}#${i}`);
      }
    });
  }

  public processNodeList(nodes: BlueNode[]): void {
    const listBlueId = BlueIdCalculator.calculateBlueIdSync(nodes);
    // Convert nodes to JSON representation for storage
    const jsonContent = nodes.map((n) => NodeToMapListOrValue.get(n));
    this.blueIdToContentMap.set(listBlueId, jsonContent);
    this.blueIdToMultipleDocumentsMap.set(listBlueId, true);
  }

  protected override fetchContentByBlueId(
    baseBlueId: string
  ): JsonBlueValue | null {
    const content = this.blueIdToContentMap.get(baseBlueId);
    const isMultipleDocuments =
      this.blueIdToMultipleDocumentsMap.get(baseBlueId);

    if (content !== undefined && isMultipleDocuments !== undefined) {
      return NodeContentHandler.resolveThisReferences(
        content,
        baseBlueId,
        isMultipleDocuments
      );
    }
    return null;
  }

  /**
   * Add single nodes to the provider
   * @param nodes - The nodes to add
   */
  public addSingleNodes(...nodes: BlueNode[]): void {
    nodes.forEach((node) => this.processNode(node));
  }

  /**
   * Add single documents (as YAML strings) to the provider
   * @param docs - The YAML documents to add
   */
  public addSingleDocs(...docs: string[]): void {
    docs.forEach((doc) => {
      const parsed = yamlBlueParse(doc);
      if (parsed !== undefined) {
        const node = NodeDeserializer.deserialize(parsed);
        this.processNode(node);
      }
    });
  }

  /**
   * Get Blue ID by name
   * @param name - The name to look up
   * @returns The Blue ID for the given name
   * @throws Error if no node with the given name exists
   */
  public getBlueIdByName(name: string): string {
    const blueIds = this.nameToBlueIdsMap.get(name);
    if (!blueIds || blueIds.length === 0) {
      throw new Error(`No node with name "${name}"`);
    }
    return blueIds[0];
  }

  /**
   * Get node by name
   * @param name - The name to look up
   * @returns The node with the given name
   * @throws Error if no node with the given name exists
   */
  public getNodeByName(name: string): BlueNode {
    const node = this.findNodeByName(name);
    if (!node) {
      throw new Error(`No node with name "${name}"`);
    }
    return node;
  }

  /**
   * Add a list and its items to the provider
   * @param list - The list of nodes to add
   */
  private addListAndItsItemsFromNodes(list: BlueNode[]): void {
    this.processNodeList(list);
    list.forEach((node) => this.processNode(node));
  }

  /**
   * Add a list and its items from a YAML document
   * @param doc - The YAML document containing a list
   */
  private addListAndItsItemsFromDoc(doc: string): void {
    const parsed = yamlBlueParse(doc);
    if (parsed !== undefined) {
      const node = NodeDeserializer.deserialize(parsed);
      const items = node.getItems();
      if (items) {
        this.addListAndItsItems(items);
      }
    }
  }

  /**
   * Add a list and its items to the provider
   * @param listOrDoc - The list of nodes or the YAML document containing a list
   */
  public addListAndItsItems(listOrDoc: BlueNode[] | string): void {
    if (Array.isArray(listOrDoc)) {
      this.addListAndItsItemsFromNodes(listOrDoc);
    } else {
      this.addListAndItsItemsFromDoc(listOrDoc);
    }
  }

  /**
   * Add a list (without processing individual items)
   * @param list - The list of nodes to add
   */
  public addList(list: BlueNode[]): void {
    this.processNodeList(list);
  }
}

File: references/java-sdk/src/main/java/blue/language/provider/BasicNodeProvider.java
package blue.language.provider;

import blue.language.Blue;
import blue.language.blueid.BlueIdCalculator;
import blue.language.model.Node;
import blue.language.preprocess.Preprocessor;
import blue.language.snapshot.ResolvedSnapshot;
import blue.language.utils.Nodes;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;

import java.util.*;
import java.util.function.Function;
import java.util.stream.IntStream;

import static blue.language.utils.UncheckedObjectMapper.JSON_MAPPER;
import static blue.language.utils.UncheckedObjectMapper.YAML_MAPPER;

public class BasicNodeProvider extends PreloadedNodeProvider {

    private Map<String, JsonNode> blueIdToContentMap;
    private Map<String, Boolean> blueIdToMultipleDocumentsMap;
    private Map<String, ResolvedSnapshot> semanticBlueIdToSnapshotMap;
    private Function<Node, Node> preprocessor;

    public BasicNodeProvider(Node... nodes) {
        this(Arrays.asList(nodes));
    }

    public BasicNodeProvider(Collection<Node> nodes) {
        this.blueIdToContentMap = new HashMap<>();
        this.blueIdToMultipleDocumentsMap = new HashMap<>();
        this.semanticBlueIdToSnapshotMap = new HashMap<>();

        Preprocessor defaultPreprocessor = new Preprocessor(this);
        this.preprocessor = defaultPreprocessor::preprocessWithDefaultBlue;

        nodes.forEach(this::processNode);
    }

    private void processNode(Node node) {
        if (Nodes.hasItemsOnly(node)) {
            processNodeWithItems(node);
        } else {
            processSingleNode(node);
        }
    }

    private void processSingleNode(Node node) {
        NodeContentHandler.ParsedContent parsedContent = NodeContentHandler.parseAndCalculateBlueId(node, preprocessor);
        blueIdToContentMap.put(parsedContent.blueId, parsedContent.content);
        blueIdToMultipleDocumentsMap.put(parsedContent.blueId, parsedContent.isMultipleDocuments);
        addToNameMap(node.getName(), parsedContent.blueId);
    }

    private void processNodeWithItems(Node node) {
        List<Node> items = node.getItems();
        processNodeList(items);

        NodeContentHandler.ParsedContent parsedContent = NodeContentHandler.parseAndCalculateBlueId(items, preprocessor);
        blueIdToContentMap.put(parsedContent.blueId, parsedContent.content);
        blueIdToMultipleDocumentsMap.put(parsedContent.blueId, true);

        IntStream.range(0, items.size()).forEach(i -> {
            Node item = items.get(i);
            addToNameMap(item.getName(), parsedContent.blueId + "#" + i);
        });
    }

    public void processNodeList(List<Node> nodes) {
        String listBlueId = BlueIdCalculator.calculateSemanticBlueId(nodes);
        JsonNode listContent = JSON_MAPPER.valueToTree(nodes);
        blueIdToContentMap.put(listBlueId, listContent);
        blueIdToMultipleDocumentsMap.put(listBlueId, true);
    }

    @Override
    protected JsonNode fetchContentByBlueId(String baseBlueId) {
        JsonNode content = blueIdToContentMap.get(baseBlueId);
        Boolean isMultipleDocuments = blueIdToMultipleDocumentsMap.get(baseBlueId);
        if (content != null && isMultipleDocuments != null) {
            return NodeContentHandler.resolveThisReferences(content, baseBlueId, isMultipleDocuments);
        }
        return null;
    }

    public void addSingleNodes(Node... nodes) {
        Arrays.stream(nodes).forEach(this::processNode);
    }

    public void addSingleDocs(String... docs) {
        Arrays.stream(docs)
                .map(doc -> YAML_MAPPER.readValue(doc, Node.class))
                .forEach(this::processNode);
    }

    public String addSingleDocsSemantic(String yaml) {
        Blue blue = new Blue(this);
        Node authoring = YAML_MAPPER.readValue(yaml, Node.class);
        ResolvedSnapshot snapshot = blue.resolveToSnapshot(authoring);
        storeSemanticSnapshot(snapshot, false);
        addNodeToNameMap(JSON_MAPPER.valueToTree(snapshot.resolvedRoot().toNode()), snapshot.rootBlueId());
        return snapshot.rootBlueId();
    }

    public String addMultipleDocsSemantic(String yaml) {
        JsonNode jsonNode = YAML_MAPPER.readTree(yaml);
        if (!jsonNode.isArray()) {
            return addSingleDocsSemantic(yaml);
        }

        Blue blue = new Blue(this);
        ArrayNode canonicalNodes = JSON_MAPPER.createArrayNode();
        List<Node> canonicalNodeList = new ArrayList<Node>();
        List<JsonNode> resolvedNodeJsons = new ArrayList<JsonNode>();

        for (int i = 0; i < jsonNode.size(); i++) {
            Node authoring = JSON_MAPPER.convertValue(jsonNode.get(i), Node.class);
            ResolvedSnapshot snapshot = blue.resolveToSnapshot(authoring);
            storeSemanticSnapshot(snapshot, false);
            resolvedNodeJsons.add(JSON_MAPPER.valueToTree(snapshot.resolvedRoot().toNode()));

            Node canonicalNode = snapshot.canonicalRoot().toNode();
            canonicalNodeList.add(canonicalNode);
            canonicalNodes.add(JSON_MAPPER.valueToTree(canonicalNode));
        }

        String listSemanticBlueId = BlueIdCalculator.calculateSemanticBlueId(canonicalNodeList);
        blueIdToContentMap.put(listSemanticBlueId, canonicalNodes);
        blueIdToMultipleDocumentsMap.put(listSemanticBlueId, true);
        for (int i = 0; i < resolvedNodeJsons.size(); i++) {
            addNodeToNameMap(resolvedNodeJsons.get(i), listSemanticBlueId + "#" + i);
        }
        return listSemanticBlueId;
    }

    public String getBlueIdByName(String name) {
        return nameToBlueIdsMap.get(name).get(0);
    }

    public Node getNodeByName(String name) {
        return findNodeByName(name).orElseThrow(() -> new IllegalArgumentException("No node with name \"" + name + "\""));
    }

    public void addListAndItsItems(List<Node> list) {
        processNodeList(list);
        list.forEach(this::processNode);
    }

    public void addListAndItsItems(String doc) {
        Node listNode = YAML_MAPPER.readValue(doc, Node.class);
        addListAndItsItems(listNode.getItems());
    }

    public void addList(List<Node> list) {
        processNodeList(list);
    }

    public Optional<ResolvedSnapshot> findSnapshotBySemanticBlueId(String semanticBlueId) {
        return Optional.ofNullable(semanticBlueIdToSnapshotMap.get(semanticBlueId));
    }

    private void storeParsedContent(NodeContentHandler.ParsedContent parsedContent) {
        blueIdToContentMap.put(parsedContent.blueId, parsedContent.content);
        blueIdToMultipleDocumentsMap.put(parsedContent.blueId, parsedContent.isMultipleDocuments);
    }

    private void addNodeToNameMap(JsonNode node, String blueId) {
        JsonNode nameNode = node.get("name");
        if (nameNode != null && !nameNode.isNull()) {
            addToNameMap(nameNode.asText(), blueId);
        }
    }

    private void storeSemanticSnapshot(ResolvedSnapshot snapshot, boolean isMultipleDocuments) {
        String semanticBlueId = snapshot.rootBlueId();
        JsonNode canonicalJson = JSON_MAPPER.valueToTree(snapshot.canonicalRoot().toNode());
        blueIdToContentMap.put(semanticBlueId, canonicalJson);
        blueIdToMultipleDocumentsMap.put(semanticBlueId, isMultipleDocuments);
        semanticBlueIdToSnapshotMap.put(semanticBlueId, snapshot);
    }
}